web3.js/test/connection.test.ts

```diff
@@ -903,6 +925,132 @@ describe('Connection', function () {
+
+      await mockRpcMessage({
+        method: 'signatureSubscribe',
+        params: [newSignature, {commitment: 'finalized'}],
+        result: new Promise(() => {}),
+      });
+      const timeoutPromise = connection.confirmTransaction(newSignature);
+
+      clock.tick(60 * 1000);
+
+      await expect(timeoutPromise).to.be.rejectedWith(
+        `Signature ${newSignature} has expired: timeout exceeded.`,
+      );
+    });
+  });
+
+  it('confirm transaction - block height exceeded', async () => {
+    const blockHeightSignature =
+      '4oCEqwGrMdBeMxpzuWiukCYqSfV4DsSKXSiVVCh1iJ6pS772X7y219JZP3mgqBz5PhsvprpKyhzChjYc3VSBQXzG';
+
+    const txBlockHeightExpired = new TransactionExpiredBlockheightExceededError(
+      blockHeightSignature,
+    );
+
+    await mockRpcMessage({
+      method: 'signatureSubscribe',
+      params: [blockHeightSignature, {commitment: 'finalized'}],
+      result: new Promise(() => {}),
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [{commitment: 'finalized'}],
+      value: 2,
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [{commitment: 'finalized'}],
+      value: 3,
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [],
+      value: 6,
+    });
+
+    await mockRpcMessage({
+      method: 'signatureSubscribe',
+      params: [blockHeightSignature, {commitment: 'finalized'}],
+      result: {err: txBlockHeightExpired},
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [],
+      error: txBlockHeightExpired,
+    });
+
+    const blockHeightPromise = connection.confirmTransaction({
+      signature: blockHeightSignature,
+      blockhash: 'sampleBlockhash',
+      lastValidBlockHeight: 5,
+    });
+
+    await expect(blockHeightPromise).to.be.rejectedWith(
+      `Signature ${blockHeightSignature} has expired: block height exceeded.`,
+    );
+  });
+
+  it('confirm transaction - block height confirmed', async () => {
+    const blockHeightSignature =
+      'LPJ18iiyfz3G1LpNNbcBnBtaS4dVBdPHKrnELqikjER2DcvB4iyTgz43nKQJH3JQAJHuZdM1xVh5Cnc5Hc7LrqC';
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [{commitment: 'processed'}],
+      value: 2,
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [{commitment: 'processed'}],
+      value: 3,
+    });
+
+    await mockRpcResponse({
+      method: 'getBlockHeight',
+      params: [{commitment: 'processed'}],
+      value: 4,
+    });
+
+    await mockRpcMessage({
+      method: 'signatureSubscribe',
+      params: [blockHeightSignature, {commitment: 'finalized'}],
+      result: Promise.resolve({err: null}),
+    });
+
+    const blockHeightPromise = connection.confirmTransaction({
+      signature: blockHeightSignature,
+      blockhash: 'sampleBlockhash',
+      lastValidBlockHeight: 6,
+    });
```

Comment:
Almost! The idea is that these things have to happen in this order in your test:

* you queue up an unresolved promise for the `signatureSubscribe` response
* you queue up some responses for `getBlockHeight`
* you make the `confirmTransaction` call
* you advance time so that every one of the `getBlockHeight` calls gets made
* _then_ you resolve the `signatureSubscribe` promise

That would be a true simulation of receiving a few blockheights before and then receiving a response in the nick of time.